### PR TITLE
`Modal`: add `__experimentalHideHeader` prop, wrap in `forwardref`

### PR DESCRIPTION
**Pull request by ciampo**
## Description

This PR contains a few changes to the `Modal` component that are required by #34153 

- add the `__experimentalHideHeader` prop, which hides the header part of the modal (title, icon, close button) and adjust styles accordingly
- wrap the component in `forwardRef`, to allow external component to get a ref of the Modal's container
- add the component (and its dependencies) to the TS config. The component is still excluded from TS checks thanks to the `//@ts-nocheck` comment at the start of each file (these comments will be removed when the component gets refactored to TypeScript)

(**Reviewing commit-by-commit should be an easier task**)

## How has this been tested?

- The component should work and behave as before across Storybook and the repo.
- The `Modal` Storybook example has been updated to showcase the `__experimentalHideHeader` prop in the "knobs" tab
- Added an additional unit test

## Screenshots 

https://user-images.githubusercontent.com/1083581/143286389-9d132339-6d27-418d-9902-6d21d2e5e3b6.mp4

## Types of changes

New feature

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/components/src/modal/index.js [hunk 1]
```diff
@@ -30,31 +37,34 @@ import StyleProvider from '../style-provider';
...
+		icon,
+		closeButtonLabel,
+		children,
+		style,
+		overlayClassName,
+		className,
+		contentLabel,
+		onKeyDown,
+		isFullScreen = false,
+		__experimentalHideHeader = false,
```
Comment (fullofcaffeine) [inline packages/components/src/modal/index.js:65]: Oh, interesting! I'll keep this convention in mind for future additions.

**Review by fullofcaffeine — APPROVED** [review]
Code looks good, storybook stories too, and all tests are passing. I think this is good to go!

File: packages/components/CHANGELOG.md [hunk 2]
```diff
@@ -13,11 +13,16 @@
 -   Fixed inconsistent padding in `UnitControl` ([#35646](https://github.com/WordPress/gutenberg/pull/35646)).
 -   Added support for RTL behavior for the `ZStack`'s `offset` prop ([#36769](https://github.com/WordPress/gutenberg/pull/36769))
 -   Fixed race conditions causing conditionally displayed `ToolsPanelItem` components to be erroneously deregistered ([36588](https://github.com/WordPress/gutenberg/pull/36588)).
+- Added `__experimentalHideHeader` prop to `Modal` component ([#36831](https://github.com/WordPress/gutenberg/pull/36831)).
```
Comment (tyxla) [inline packages/components/CHANGELOG.md]: I'm not sure there's a more minor nit ever, but we seem to be aligning these to the title vertically:

```suggestion
-   Added `__experimentalHideHeader` prop to `Modal` component ([#36831](https://github.com/WordPress/gutenberg/pull/36831)).
```

File: packages/components/src/modal/test/index.js [hunk 3]
```diff
@@ -46,4 +46,15 @@ describe( 'Modal', () => {
 		const titleId = within( dialog ).getByText( 'Test Title' ).id;
 		expect( dialog ).toHaveAttribute( 'aria-labelledby', titleId );
 	} );
+
+	it( 'hides the header when the `__experimentalHideHeader` prop is used', () => {
+		render(
+			<Modal title="Test Title" __experimentalHideHeader={ true }>
```
Comment (tyxla) [inline packages/components/src/modal/test/index.js]: Seems like we could ditch the `={ true }`:

```suggestion
			<Modal title="Test Title" __experimentalHideHeader>
```

**Review by tyxla — APPROVED** [review]
Looks good and tests well 🚀

File: packages/components/CHANGELOG.md (hunk 2 above)
Comment (ciampo) [reply packages/components/CHANGELOG.md]: Minor nits are the best, especially when they come with a good code suggestion!

**Comment by ciampo** [comment]
Thank you @fullofcaffeine and @tyxla for the quick reviews 🙇

**Pull request merged**